pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package compactor
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"path"
+
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+
+	"github.com/grafana/mimir/pkg/storage/bucket"
+)
+
+// CreateBlockUpload requests the creation of a block upload session.
+func (c *MultitenantCompactor) CreateBlockUpload(ctx context.Context, tenantID, blockID string) error {
+	level.Info(c.logger).Log("msg", "creating block upload session", "user", tenantID, "block_id", blockID)
+	// TODO: Verify that block hasn't already been ingested
+	return nil
+}
+
+// UploadBlockFile uploads a block file.
+func (c *MultitenantCompactor) UploadBlockFile(ctx context.Context, tenantID, blockID, pth string, r *http.Request) error {
+	if path.Base(pth) == "meta.json" {
+		return fmt.Errorf("meta.json is not allowed")
+	}
+
+	level.Info(c.logger).Log("msg", "uploading block file", "user", tenantID,
+		"block_id", blockID, "path", pth, "size", r.ContentLength)
+
+	dst := path.Join(blockID, pth)
```

Comment:
What would happen here if path is `../othertenant/maliciousblock/chunk`?

Reply:
@colega does `..` work in object store paths?

Reply:
You're right, I checked and it _doesn't seem_ that it would work on S3 or GCS, however I think that it's worth validating here instead of blindly relying on the backend (btw it would work on filesystem backend)

Reply:
Note that `path.Join` would handle `..`, so `dst` would not have it anymore, and would point to incorrect location.

Reply:
Agree we should think this over when going over the details of the PR.

Reply:
@colega @pstibrany I added some normalization/validation logic with `path.Clean`, can you please have a look and see if it makes sense?

Reply:
See [other discussion](https://github.com/grafana/mimir/pull/1694#discussion_r877552512).